spec/validator.ml

```diff
@@ -1,180 +1,111 @@
+  let encode (offset, nonces) =
+    Abi.encode (`Uint256 offset :: FROST.encode_nonces nonces)
+end)
 
-  let init (_ : Address.t list) : t = failwith "place holder"
-  let remove (_ : t) (_ : 'a) : t = failwith "place holder"
-  let remove_missing (_ : t) (_ : 'a) : t = failwith "place holder"
-  let cardinal (_ : t) : int = failwith "place holder"
-  let identifier (_ : t) (_ : Address.t) : int = failwith "place holder"
-  let addr (_ : t) (_ : int) : Address.t = failwith "place holder"
-end
+module VerificationMTree = MerkleTree.Make (struct
+  type t =
+    FROST.(
+      Identifier.t * share_commitment * lagrange_coefficient * group_commitment)
 
-type group_id = string
+  let encode (id, ri, l, r) =
+    Abi.encode @@ FROST.encode_verification_share id ri l r
+end)
 
-module FROSTCoordinator = struct
-  let group_id (_ : ParticipantSet.t) (_ : int) (_ : int) (_ : string) :
-      group_id =
-    failwith "place holder"
-end
+(** A marker type to indicate state that is local to a specific validator and
+    only known by them. *)
+type 'a local = Local of 'a
 
 type group = {
-  id : group_id;  (** The set of participants in the group. *)
+  id : GroupId.t;
   participants : ParticipantSet.t;
-      (** The identifier of the validator in the group. *)
-  key : Secp256k1.public_key;  (** The participant share ID. *)
-  share : Secp256k1.secret_key;
+  key : FROST.verification_key;
+  me : FROST.Identifier.t local;
+  share : FROST.signing_share local;
 }
 
-type epoch = { id : int; group : group }
+type epoch = { epoch : int; group : group }
 
 type rollover =
   | Collecting_key_gen_commitments of {
       epoch : int;
-      id : group_id;
-      coefficients : Secp256k1.scalar list;
-      commitments : Secp256k1.point list IntMap.t;
-      participants : ParticipantSet.t;
+      group : GroupId.t;
+      participants : AddressSet.t;
+      coefficients : FROST.KeyGen.coefficients local;
+      commitments : FROST.KeyGen.commitments ParticipantMap.t;
       deadline : int;
     }
   | Collecting_key_gen_secret_shares of {
       epoch : int;
-      id : group_id;
-      key : Secp256k1.point;
-      commitments : Secp256k1.point list IntMap.t;
-      shares : Secp256k1.scalar IntMap.t;
-      last_participant : Address.t;
-      participants : ParticipantSet.t;
+      group : GroupId.t;
+      participants : AddressSet.t;
+      key : FROST.verification_key;
+      coefficients : FROST.KeyGen.coefficients local;
+      commitments : FROST.KeyGen.commitments ParticipantMap.t;
+      shares : FROST.KeyGen.secret_share ParticipantMap.t local;
+      last_participant : FROST.Identifier.t;
       deadline : int;
     }
   | Signing_epoch_rollover of { epoch : epoch; message : string }
-  | Staged_epoch of epoch
+  | Staged_epoch of { epoch : epoch }
+  | Skipped of { epoch : int }
 
 type preprocess_group = {
-  nonces : FROST.nonces IntMap.t;
-  pending : FROST.nonces list StringMap.t;
+  nonces : (FROST.nonces * MerkleTree.proof) IntMap.t;
```

Comment:
what does the *  do here and in the next line? What would be the corresponding ts type?